src/algorithms/tonal/crepe.cpp

```diff
@@ -0,0 +1,131 @@
+
+const char* CREPE::name = "CREPE";
+const char* CREPE::category = "Pitch";
+const char* CREPE::description = DOC(
+  "This algorithm estimates pitch using CREPE models.\n"
+  "\n"
+  "This algorithm is a wrapper to post-process the activations generated by TensorflowPredictCREPE. "
+  "`time` contains the timestamps in which the pitch was estimated. `frequency` is the vector of "
+  "pitch estimations in Hz. `confidence` expresses the confidence in the presence of pitch for each "
+  "timestamp as value between 0 to 1. `activations` is a time by sigmoid activations matrix returned "
+  " by the neural network.\n"
+  "\n"
+  "See TensorflowPredictCREPE for details about the rest of parameters.\n"
+  "The recommended pipeline is as follows::\n"
+  "\n"
+  "  MonoLoader(sampleRate=16000) >> CREPE()\n"
+  "\n"
+  "Note: This algorithm does not make any check on the input model so it is "
+  "the user's responsibility to make sure it is a valid one.\n"
+  "\n"
+  "References:\n"
+  "\n"
+  "1. CREPE: A Convolutional Representation for Pitch Estimation. "
+  "Jong Wook Kim, Justin Salamon, Peter Li, Juan Pablo Bello. "
+  "Proceedings of the IEEE International Conference on Acoustics, Speech, and Signal "
+  "Processing (ICASSP), 2018.\n"
+  "\n"
+  "2. Original models and code at https://github.com/marl/crepe/\n"
+  "\n"
+  "3. Supported models at https://essentia.upf.edu/models/\n\n");
+
+
+void CREPE::configure() {
+  _tensorflowPredictCREPE->configure(INHERIT("graphFilename"),
+                                     INHERIT("input"),
+                                     INHERIT("output"),
+                                     INHERIT("hopSize"),
+                                     INHERIT("batchSize"));
+
+  _hopSize = parameter("hopSize").toFloat();
+  // _viterbi = parameter("viterbi").toBool();
+
+  for (int i = 0; i < _nPitches - 1; ++i) {
+    _centsMapping.push_back(_delta * i + _shift);
+  }
+  _centsMapping.push_back(_end + _shift);
+}
+
+
+void CREPE::compute() {
+  const vector<Real>& audio = _audio.get();
+  vector<Real>& time = _time.get();
+  vector<Real>& frequency = _frequency.get();
+  vector<Real>& confidence = _confidence.get();
+  vector<vector<Real> >& activations = _activations.get();
+
+  _tensorflowPredictCREPE->input("signal").set(audio);
+  _tensorflowPredictCREPE->output("predictions").set(activations);
+  _tensorflowPredictCREPE->compute();
+
+  int timestamps = activations.size();
+
+  time.resize(timestamps);
+  frequency.resize(timestamps);
+  confidence.resize(timestamps);
+  
+  frequency.assign(timestamps, 0);
+
+  // Activations to cents.
+  vector<Real> cents = toLocalAverageCents(activations);
+
+  for (int i = 0; i < timestamps; i++) {
+    // Get the timestamp of each pitch prediction.
+    time[i] = i * _hopSize / 1000.0;
+
+    // Get the confidences (max Activation per timestamp).
+    confidence[i] = activations[i][argmax(activations[i])];
+
+    // Cents to frequencies.
+    if (!isnan(cents[i])) {
```

Comment:
`nan` can happen when:
- a frame is unvoiced, so all the activations sum to zero and there is a division by zero computing the cents.
- the activation matrix itself contain `nans`. I guess this could happen, but I have never seen it.

I think it's better to handle the unvoiced timestamps in the Herz level because:
1. while we can use 0Hz to represent unvoiced timestamps, 0 cents actually refers to the lowest pitch the model is able to predict.
2. This implementation is parallel to the original Python implementation, which can be useful for future debugging.

Reply:
A new observation:
In the [original implementation](https://github.com/marl/crepe/blob/a666a03011a9cdab70e9abd0b5009ad60c5f8926/crepe/core.py#L207), frames are normalized in a way that constant frames (e.g., zeros) produce `NaNs`: `( (x - mean) / stddev )`

These `NaNs` are propagated to the activation matrix, and are then interpreted as frequencies equal to 0 according to the logic that we implemented on line 104.

CREPE has an open [PR](https://github.com/marl/crepe/pull/76/) proposing to use `EPS` to prevent division by 0, but the authors haven't made any comment about it.
I have tried the PR, and it effectively prevents the `NaNs` to happen and the CNN return incorrect pitch values with low confidence (activation) values. I guess the authors may prefer the current behavior, as it returns clean zeros for constant frames. 

Essentia has two mechanisms to avoid this:
- By default `FrameCutter` adds noise to silence frames.
- `TensorNormalice` does not divide by the stddev when it is equal to zero

For now, I decided to deactivate both mechanisms so that our normalization also returns `NaNs` on constant frames, so that we mimic the original implementation.

Reply:
On the other side, `NaN` is not part of JSON and we need to double-check if it is possible to store such an output with our `YamlOutput` algorithm (and we are missing a unit test for that in `test_yamloutput.py`).